packages/datadog-instrumentations/src/helpers/router-helper.js

```diff
@@ -0,0 +1,176 @@
+      continue
+    }
+
+    const normalized = normalizeRoutePath(current)
+
+    if (normalized !== null) {
+      paths.push(normalized)
+    }
+  }
+
+  return paths
+}
+
+function setRouterMountPath (router, mountPath) {
+  if (!router || typeof mountPath !== 'string') return
+  const existing = routerMountPaths.get(router)
+  if (existing) {
+    existing.add(mountPath)
+  } else {
+    routerMountPaths.set(router, new Set([mountPath]))
+  }
+}
+
+function getRouterMountPaths (router) {
+  const paths = routerMountPaths.get(router)
+  if (!paths) return []
+  return [...paths]
+}
+
+function setLayerMatchers (layer, matchers) {
+  layerMatchers.set(layer, matchers)
+}
+
+function getLayerMatchers (layer) {
+  return layerMatchers.get(layer)
+}
+
+function normalizeMethodName (method) {
+  return method === '_all' || method === 'all' ? '*' : method
+}
+
+function getRouteFullPaths (route, prefix) {
+  if (!route) return []
+
+  const routePaths = normalizeRoutePaths(route.path)
+  const pathsToPublish = routePaths.length ? routePaths : ['']
+
+  return pathsToPublish.map(routePath => joinPath(prefix, routePath))
+}
+
+function markAppMounted (router) {
+  if (router) appMountedRouters.add(router)
+}
+
+function isAppMounted (router) {
+  return appMountedRouters.has(router)
+}
+
+/**
+ * Normalise the optional mount argument provided to app.use()/router.use().
+ * Express accepts strings, regex, arrays (possibly nested), or
+ * no mount path at all; this helper returns the flattened set of paths along
+ * with the index where actual middleware arguments start.
+ */
+function extractMountPaths (args) {
+  const firstArg = args[0]
+  const hasMount = typeof firstArg === 'string' || firstArg instanceof RegExp || Array.isArray(firstArg)
+
+  if (!hasMount) {
+    return { mountPaths: ['/'], startIdx: 0 }
+  }
+
+  const paths = normalizeRoutePaths(firstArg)
+  return {
+    mountPaths: paths.length ? paths : ['/'],
+    startIdx: 1
+  }
+}
+
+/**
+ * Detect cycle router graphs.
+ */
+function hasRouterCycle (router, stack = new Set()) {
+  if (!router?.stack?.length) return false
+  if (stack.has(router)) return true
+
+  stack.add(router)
+
+  for (const layer of router.stack) {
+    if (!layer?.route && layer?.handle?.stack?.length) {
+      const hasCycle = hasRouterCycle(layer.handle, stack)
+      if (hasCycle) {
+        stack.delete(router)
+        return true
+      }
+    }
+  }
+
+  stack.delete(router)
+  return false
+}
+
+function wrapRouteMethodsAndPublish (route, paths, publish) {
+  if (!route || !paths?.length || typeof publish !== 'function') return
+
+  const uniquePaths = [...new Set(paths.filter(Boolean))]
+  if (!uniquePaths.length) return
+
+  METHODS.forEach(method => {
+    if (typeof route[method] !== 'function') return
+
+    shimmer.wrap(route, method, (originalMethod) => function wrappedRouteMethod () {
+      for (const path of uniquePaths) {
+        publish({
+          method: normalizeMethodName(method),
```

Comment:
`method` is constant, so no need to keep normalizing it:

```suggestion
    const normalizedMethod = normalizeMethodName(method)

    shimmer.wrap(route, method, (originalMethod) => function wrappedRouteMethod () {
      for (const path of uniquePaths) {
        publish({
          method: normalizedMethod,
```

Alternatively, if we think it's not likely that the wrapped method is called, we could move it into the body of it to make the normalization lazy, like so:

```suggestion

    shimmer.wrap(route, method, (originalMethod) => function wrappedRouteMethod () {
      const normalizedMethod = normalizeMethodName(method)
      for (const path of uniquePaths) {
        publish({
          method: normalizedMethod,
```